torch/csrc/jit/codegen/cuda/kernel_cache.cpp

```diff
@@ -8,14 +11,542 @@ namespace jit {
+          TORCH_INTERNAL_ASSERT(
+              parsing_graph->outputs().size() == 1 &&
+                  parsing_graph->outputs()[0] == n->output(),
+              "supporfor graph with reduction is limited to single output from reduction node");
+          auto dims_list = constant_as<c10::List<int64_t>>(n->input(1));
+          TORCH_INTERNAL_ASSERT(
+              dims_list.has_value(), "reduction axes should be constant");
+          std::vector<int64_t> adjusted_reduction_axes;
+          for (const auto dim : dims_list->vec()) {
+            // adjust reduction axis to be the permuted axis;
+            for (size_t j = 0; j < input_permutation.size(); j++) {
+              // follow the permutation to resolve the new reduction axes;
+              if (input_permutation[j] == dim) {
+                adjusted_reduction_axes.emplace_back(j);
+                break;
+              }
+            }
+          }
+          parsing_graph->setInsertPoint(n);
+          auto const_ival_axes =
+              parsing_graph->insertConstant(IValue(adjusted_reduction_axes));
+          n->replaceInput(1, const_ival_axes);
+        }
+      }
+    }
+  }
+
   TORCH_INTERNAL_ASSERT(
-      entry == nullptr,
-      "At this time FusionExecutorCache only supports one entry.");
-  entry = new FusionExecutor();
-  entry->compileFusion(fusion, options);
+      input_stacks_.back().device_.has_value(),
+      "device is not set for fusion executor, something went wrong in NvFuser");
+  fec_cache_.emplace_back(std::make_unique<FusionExecutorCache>(
+      parseJitIR(parsing_graph), input_stacks_.back().device_.value()));
+  return fec_cache_.back().get();
+}
+
+GraphCache::GraphCache(std::shared_ptr<Graph> graph)
+    : graph_(std::move(graph)) {
+  // [ NOTE - reduction in graph ]
+  //
+  // reduction complicates our permutation in integration, it addes two things:
+  // 1. we need to adjust output_permutation_;
+  //    because of dimension elimination during permutation (not necessarily,
+  //    given the `keepdim` argument.) this needs to be accommodated later when
+  //    we added the support.
+  // 2. adjust reduction axes for the permutation;
+  //    permute changes the semantics of axes, we need to update the reduction
+  //    axes in the graph in order to match the behavior;
+  reduction_axes_ = graphReductionAxes(graph_);
+
+  // compile a kernel if we have enough information from graph (profiling
+  // record)
+  if (IsNewExecutorEnabled()) {
+    createFusionExecutorCache(
+        InputsRequirement(graph_, toVector(reduction_axes_)));
+  }
+}
+
+std::vector<at::Tensor> GraphCache::runGraphWithInputs(
+    const at::ArrayRef<IValue>& inputs) {
+  InputsRequirement input_stack(inputs, toVector(reduction_axes_));
+  FusionExecutorCache* fusion_executor_cache = nullptr;
+
+  // TODO: hash indexing;
+  for (int i = 0; i < static_cast<int>(fec_cache_.size()); i++) {
+    if (input_stack.complyWith(input_stacks_[i])) {
```

Comment:
Does this at least return relatively quickly? Checking most common things or quickest things to check that likely don't match?

Reply:
Nope it doesn't. There's really not a good way I can think about that would make this check fast.
Hence I'm thinking about just adding a hash/encode of input properties, so we can map seen inputs directly to the cached fusion executor.